Implementation of OutputPin requires InputPin

Hey, me again. So the SPI bus seems to be working, but now I'm realizing that I can't use `.forward()` to convert my embedded hal 2.x output pins into embedded hal 1.0.0-alpha.y versions. It appears that the implementation requires the pin to be both an input and an output for some reason. Possibly a bug?

https://github.com/ryankurte/embedded-hal-compat/blob/2311e03e3304e3f76c09dc52bbd1cb1ad98875c8/src/forward.rs#L92-L94

```
error[E0277]: the trait bound `stm32l4xx_hal::gpio::Pin<Output<PushPull>, stm32l4xx_hal::gpio::L8, 'A', 4>: _embedded_hal_digital_InputPin` is not satisfied
  --> src/lib.rs:81:25
   |
81 |             trait_check(&reader_cs.forward());
   |             ----------- ^^^^^^^^^^^^^^^^^^^^ the trait `_embedded_hal_digital_InputPin` is not implemented for `stm32l4xx_hal::gpio::Pin<Output<PushPull>, stm32l4xx_hal::gpio::L8, 'A', 4>`
   |             |
   |             required by a bound introduced by this call
   |
   = help: the trait `_embedded_hal_digital_InputPin` is implemented for `OldInputPin<T>`
   = note: required for `stm32l4xx_hal::gpio::Pin<Output<PushPull>, stm32l4xx_hal::gpio::L8, 'A', 4>` to implement `stm32l4xx_hal::prelude::InputPin`
   = note: required for `Forward<stm32l4xx_hal::gpio::Pin<Output<PushPull>, stm32l4xx_hal::gpio::L8, 'A', 4>>` to implement `embedded_hal_compat::embedded_hal::digital::OutputPin`
```

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -1,5 +1,10 @@
 name: tests
-on: [push, pull_request]
+on:
+  push:
+    branches: [ main ]
+    tags: [ 'v*' ]
+  pull_request:
+    branches: [ main ]
 
 env:
   RUSTFLAGS: '--deny warnings'

--- a/src/forward.rs
+++ b/src/forward.rs
@@ -2,31 +2,39 @@
 //! A compatibility layer to alleviate (some) of the issues resolving from changes to embedded-hal
 // Copyright 2021 Ryan Kurte
 
+use core::marker::PhantomData;
+
 /// Forward compatibility container object.
 /// This is generic over different E-H types and will provide adaption
 /// depending on the bound type.
-pub struct Forward<T> {
+pub struct Forward<T, M = ()> {
     inner: T,
+    _marker: PhantomData<M>,
 }
 
 /// Helper trait to convert a type for forward compatibility
 /// call `.forward()` on `e-h@0.2.x` types to create an `e-h@1.x.x` compatible wrapper object
-pub trait ForwardCompat<T> {
-    fn forward(self) -> Forward<T>;
-}
-
-impl<T> ForwardCompat<T> for T {
+pub trait ForwardCompat<T, M = ()> {
     /// Create an e-h-c forward compatibility wrapper around and e-h object
     /// Available methods depend on the wrapped type
-    fn forward(self) -> Forward<T> {
+    fn forward(self) -> Forward<T, M>;
+}
+
+/// Blanket [ForwardCompat] implementation
+/// (note input/output/io pins may require type annotations)
+impl<T, M> ForwardCompat<T, M> for T {
+    fn forward(self) -> Forward<T, M> {
         Forward::new(self)
     }
 }
 
-impl<T> Forward<T> {
+impl<T, M> Forward<T, M> {
     /// Create a new compatibility wrapper object
-    pub fn new(inner: T) -> Forward<T> {
-        Forward { inner }
+    pub fn new(inner: T) -> Forward<T, M> {
+        Forward {
+            inner,
+            _marker: PhantomData,
+        }
     }
 
     /// Fetch a reference to the wrapped object
@@ -58,6 +66,7 @@ pub struct ForwardError<E>(pub E);
 // Digital / GPIOs
 mod digital {
     use super::{Forward, ForwardError};
+    use crate::markers::{ForwardInputPin, ForwardIoPin, ForwardOutputPin};
 
     impl<E: core::fmt::Debug> eh1_0::digital::Error for ForwardError<E> {
         fn kind(&self) -> eh1_0::digital::ErrorKind {
@@ -65,7 +74,31 @@ mod digital {
         }
     }
 
-    impl<T, E> eh1_0::digital::ErrorType for Forward<T>
+    impl<T, E> eh1_0::digital::ErrorType for Forward<T, ForwardInputPin>
+    where
+        T: eh0_2::digital::v2::InputPin<Error = E>,
+        E: core::fmt::Debug,
+    {
+        type Error = super::ForwardError<E>;
+    }
+
+    impl<T, E> eh1_0::digital::InputPin for Forward<T, ForwardInputPin>
+    where
+        T: eh0_2::digital::v2::InputPin<Error = E>,
+        E: core::fmt::Debug,
+    {
+        /// Is the input pin high?
+        fn is_high(&self) -> Result<bool, Self::Error> {
+            self.inner.is_high().map_err(ForwardError)
+        }
+
+        /// Is the input pin low?
+        fn is_low(&self) -> Result<bool, Self::Error> {
+            self.inner.is_low().map_err(ForwardError)
+        }
+    }
+
+    impl<T, E> eh1_0::digital::ErrorType for Forward<T, ForwardOutputPin>
     where
         T: eh0_2::digital::v2::OutputPin<Error = E>,
         E: core::fmt::Debug,
@@ -73,7 +106,31 @@ mod digital {
         type Error = super::ForwardError<E>;
     }
 
-    impl<T, E> eh1_0::digital::InputPin for Forward<T>
+    impl<T, E> eh1_0::digital::OutputPin for Forward<T, ForwardOutputPin>
+    where
+        T: eh0_2::digital::v2::OutputPin<Error = E>,
+        E: core::fmt::Debug,
+    {
+        /// Set the output as high
+        fn set_high(&mut self) -> Result<(), Self::Error> {
+            self.inner.set_high().map_err(ForwardError)
+        }
+
+        /// Set the output as low
+        fn set_low(&mut self) -> Result<(), Self::Error> {
+            self.inner.set_low().map_err(ForwardError)
+        }
+    }
+
+    impl<T, E> eh1_0::digital::ErrorType for Forward<T, ForwardIoPin>
+    where
+        T: eh0_2::digital::v2::OutputPin<Error = E> + eh0_2::digital::v2::InputPin<Error = E>,
+        E: core::fmt::Debug,
+    {
+        type Error = super::ForwardError<E>;
+    }
+
+    impl<T, E> eh1_0::digital::InputPin for Forward<T, ForwardIoPin>
     where
         T: eh0_2::digital::v2::InputPin<Error = E> + eh0_2::digital::v2::OutputPin<Error = E>,
         E: core::fmt::Debug,
@@ -89,9 +146,9 @@ mod digital {
         }
     }
 
-    impl<T, E> eh1_0::digital::OutputPin for Forward<T>
+    impl<T, E> eh1_0::digital::OutputPin for Forward<T, ForwardIoPin>
     where
-        T: eh0_2::digital::v2::OutputPin<Error = E>,
+        T: eh0_2::digital::v2::InputPin<Error = E> + eh0_2::digital::v2::OutputPin<Error = E>,
         E: core::fmt::Debug,
     {
         /// Set the output as high

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -17,6 +17,9 @@
 //! use with `v1.0.x` consumers, so you can drop these wrapped types into drivers expecting
 //! `v1.0.x` types.
 //!
+//! Note that GPIO pins will require annotation with marker types (see [markers]) to select
+//! input / output / combined modes.
+//!
 //!```
 //! # use core::convert::Infallible;
 //! # pub struct OutputPin0_2;
@@ -44,7 +47,7 @@
 //! #         Ok(false)
 //! #     }
 //! # }
-//! use embedded_hal_compat::ForwardCompat;
+//! use embedded_hal_compat::{Forward, ForwardCompat, markers::*};
 //!
 //! // Create e-h v0.2.x based type (mock)
 //! let mut old = OutputPin0_2;
@@ -52,7 +55,7 @@
 //! let _ = eh0_2::digital::v2::OutputPin::set_high(&mut old);
 //!
 //! // Apply forward compatibility wrapper
-//! let mut new = old.forward();
+//! let mut new: Forward<_, ForwardIoPin> = old.forward();
 //! // Access via e-h v1.x.x methods
 //! let _ = eh1_0::digital::OutputPin::set_high(&mut new);
 //!```
@@ -121,6 +124,7 @@ pub use eh0_2;
 pub use eh1_0;
 
 mod forward;
+pub mod markers;
 mod reverse;
 
 // Forward compatibility wrapper trait, access using `.forward()`

--- a/src/markers.rs
+++ b/src/markers.rs
@@ -1,0 +1,8 @@
+/// Marker for input only pins
+pub struct ForwardInputPin;
+
+/// Marker for output only pins
+pub struct ForwardOutputPin;
+
+/// Marker for input-output pins
+pub struct ForwardIoPin;

--- a/tests/digital_forward.rs
+++ b/tests/digital_forward.rs
@@ -83,7 +83,7 @@ fn io_pin_forward() {
 #[test]
 fn input_pin_forward() {
     let periph_0_2 = InputPin;
-    let mut periph_1_0 = periph_0_2.forward();
+    let periph_1_0 = periph_0_2.forward();
     assert!(eh1_0::digital::InputPin::is_high(&periph_1_0).unwrap());
 }
 

--- a/tests/digital_forward.rs
+++ b/tests/digital_forward.rs
@@ -1,20 +1,31 @@
-use embedded_hal_compat::ForwardCompat;
+use embedded_hal_compat::{markers::*, Forward, ForwardCompat};
 
 #[derive(Debug)]
-enum PinError {
+enum InputPinError {
     _Something,
 }
 
-impl eh1_0::digital::Error for PinError {
+impl eh1_0::digital::Error for InputPinError {
     fn kind(&self) -> eh1_0::digital::ErrorKind {
         eh1_0::digital::ErrorKind::Other
     }
 }
 
-struct Peripheral;
+#[derive(Debug)]
+enum OutputPinError {
+    _Something,
+}
 
-impl eh0_2::digital::v2::OutputPin for Peripheral {
-    type Error = PinError;
+impl eh1_0::digital::Error for OutputPinError {
+    fn kind(&self) -> eh1_0::digital::ErrorKind {
+        eh1_0::digital::ErrorKind::Other
+    }
+}
+
+struct IoPin;
+
+impl eh0_2::digital::v2::OutputPin for IoPin {
+    type Error = InputPinError;
 
     fn set_high(&mut self) -> Result<(), Self::Error> {
         Ok(())
@@ -24,8 +35,34 @@ impl eh0_2::digital::v2::OutputPin for Peripheral {
     }
 }
 
-impl eh0_2::digital::v2::InputPin for Peripheral {
-    type Error = PinError;
+impl eh0_2::digital::v2::InputPin for IoPin {
+    type Error = InputPinError;
+
+    fn is_high(&self) -> Result<bool, Self::Error> {
+        Ok(true)
+    }
+    fn is_low(&self) -> Result<bool, Self::Error> {
+        Ok(false)
+    }
+}
+
+struct OutputPin;
+
+impl eh0_2::digital::v2::OutputPin for OutputPin {
+    type Error = OutputPinError;
+
+    fn set_high(&mut self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+    fn set_low(&mut self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+struct InputPin;
+
+impl eh0_2::digital::v2::InputPin for InputPin {
+    type Error = InputPinError;
 
     fn is_high(&self) -> Result<bool, Self::Error> {
         Ok(true)
@@ -36,9 +73,23 @@ impl eh0_2::digital::v2::InputPin for Peripheral {
 }
 
 #[test]
-fn can_forward() {
-    let periph_0_2 = Peripheral;
-    let mut periph_1_0 = periph_0_2.forward();
+fn io_pin_forward() {
+    let periph_0_2 = IoPin;
+    let mut periph_1_0: Forward<_, ForwardIoPin> = periph_0_2.forward();
     assert!(eh1_0::digital::OutputPin::set_high(&mut periph_1_0).is_ok());
     assert!(eh1_0::digital::InputPin::is_high(&periph_1_0).unwrap());
+}
+
+#[test]
+fn input_pin_forward() {
+    let periph_0_2 = InputPin;
+    let mut periph_1_0 = periph_0_2.forward();
+    assert!(eh1_0::digital::InputPin::is_high(&periph_1_0).unwrap());
+}
+
+#[test]
+fn output_pin_forward() {
+    let periph_0_2 = OutputPin;
+    let mut periph_1_0 = periph_0_2.forward();
+    assert!(eh1_0::digital::OutputPin::set_high(&mut periph_1_0).is_ok());
 }

